pkg/ingester/client/wrapped.go

```diff
@@ -0,0 +1,432 @@
+			}
+			if postIndex > l {
+				return io.ErrUnexpectedEOF
+			}
+			m.UserId = string(dAtA[iNdEx:postIndex])
+			iNdEx = postIndex
+		case 3:
+			if wireType != 2 {
+				return fmt.Errorf("proto: wrong wireType = %d for field Labels", wireType)
+			}
+			var msglen int
+			for shift := uint(0); ; shift += 7 {
+				if shift >= 64 {
+					return ErrIntOverflowIngester
+				}
+				if iNdEx >= l {
+					return io.ErrUnexpectedEOF
+				}
+				b := dAtA[iNdEx]
+				iNdEx++
+				msglen |= int(b&0x7F) << shift
+				if b < 0x80 {
+					break
+				}
+			}
+			if msglen < 0 {
+				return ErrInvalidLengthIngester
+			}
+			postIndex := iNdEx + msglen
+			if postIndex < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if postIndex > l {
+				return io.ErrUnexpectedEOF
+			}
+			m.Labels = append(m.Labels, mimirpb.LabelAdapter{})
+			if err := m.Labels[len(m.Labels)-1].Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
+				return err
+			}
+			iNdEx = postIndex
+		case 4:
+			if wireType != 2 {
+				return fmt.Errorf("proto: wrong wireType = %d for field Chunks", wireType)
+			}
+			var msglen int
+			for shift := uint(0); ; shift += 7 {
+				if shift >= 64 {
+					return ErrIntOverflowIngester
+				}
+				if iNdEx >= l {
+					return io.ErrUnexpectedEOF
+				}
+				b := dAtA[iNdEx]
+				iNdEx++
+				msglen |= int(b&0x7F) << shift
+				if b < 0x80 {
+					break
+				}
+			}
+			if msglen < 0 {
+				return ErrInvalidLengthIngester
+			}
+			postIndex := iNdEx + msglen
+			if postIndex < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if postIndex > l {
+				return io.ErrUnexpectedEOF
+			}
+			// << NON AUTO-GENERATED CODE >>
+			//
+			//  m.Chunks = append(m.Chunks, Chunk{})
+			//  if err := m.Chunks[len(m.Chunks)-1].Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
+			//      return err
+			//  }
+			//
+			if reusedChunks < poolChunksLength {
+				if err := m.Chunks[reusedChunks].Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
+					return err
+				}
+				reusedChunks++
+			} else {
+				m.Chunks = append(m.Chunks, Chunk{})
+				if err := m.Chunks[len(m.Chunks)-1].Unmarshal(dAtA[iNdEx:postIndex]); err != nil {
+					return err
+				}
+			}
+			// << NON AUTO-GENERATED CODE >>
+			iNdEx = postIndex
+		default:
+			iNdEx = preIndex
+			skippy, err := skipIngester(dAtA[iNdEx:])
+			if err != nil {
+				return err
+			}
+			if skippy < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if (iNdEx + skippy) < 0 {
+				return ErrInvalidLengthIngester
+			}
+			if (iNdEx + skippy) > l {
+				return io.ErrUnexpectedEOF
+			}
+			iNdEx += skippy
+		}
+	}
+
+	if iNdEx > l {
+		return io.ErrUnexpectedEOF
+	}
+	return nil
+}
+
+// ReuseQueryStreamResponse puts chunk and label slices contained in qr back into a sync.Pool for reuse.
+func ReuseQueryStreamResponse(qr *QueryStreamResponse) {
+	for i := 0; i < len(qr.Chunkseries); i++ {
+		if len(qr.Chunkseries[i].Chunks) > maxInPoolChunksSliceSize || len(qr.Chunkseries[i].Labels) > maxInPoolLabelsSliceSize {
```

Comment:
> len(qr.Chunkseries[i].Labels) > maxInPoolLabelsSliceSize

I don't fully understand what you're trying to do here. `maxInPoolLabelsSliceSize` is a very big number (16k) so we're not putting back in the pool slides bigger than 16k, but we can't have series with 16k labels. We validate input series and by default we limit to 30 labels per series.

Could you help me understand what we want to achieve here?

Reply:
Intention here is trying to avoid sending big slices back to the pool, so if either labels or chunks backed arrays have a capacity bigger than a given threshold, we let it to be reclaimed by the GC.

Nevertheless, if we are fully confident that those stream responses are never going to grow beyond certain limit, I guess we could simply drop this guard. WDYT?

Reply:
Discussed offline. We can safely drop this cap check. 👍